toolchain/docs/parse.md

```diff
@@ -0,0 +1,917 @@
+    subgraph nodes["Parsed nodes"]
+        direction BT
+        node1[var]
+        node2[x]
+        node3[:]
+        node4[i32]
+        node5[=]:::moved
+    end
+
+    classDef hidden visibility:hidden,display:none
+    classDef moved fill:#0F0,color:#000
+
+    node1 ~~~~ root
+    node3 --- node2 & node4
+    node3 ~~~ root
+    node5 ~~~~ root
+```
+
+The expression is a subtree with `+` as the parent, and the two operands as
+child nodes.
+
+```mermaid
+flowchart BT
+    subgraph tokens["Remaining tokens"]
+        token1[var]:::used
+        token2[x]:::used
+        token3[:]:::used
+        token4[i32]:::used
+        token5[=]:::used
+        token6[y]:::used
+        token7[+]:::used
+        token8[1]:::used
+        token9[;]
+    end
+
+    classDef used visibility:hidden
+```
+
+```mermaid
+flowchart BT
+    root:::hidden
+    subgraph nodes["Parsed nodes"]
+        direction BT
+        node1[var]
+        node2[x]
+        node3[:]
+        node4[i32]
+        node5[=]
+        node6[y]:::moved
+        node7[+]:::moved
+        node8[1]:::moved
+    end
+
+    classDef hidden visibility:hidden,display:none
+    classDef moved fill:#0F0,color:#000
+
+    node1 ~~~~ root
+    node3 --- node2 & node4
+    node3 ~~~ root
+    node5 ~~~~ root
+    node7 --- node6 & node8
+    node7 ~~~ root
+```
+
+Finally, the `;` is used as the "root" of the variable declaration. It's
+explicitly tracked as the `;` for a variable declaration so that it's
+unambiguously bracketed by `var`.
+
+```mermaid
+flowchart BT
+    root:::hidden
+    subgraph nodes["Parsed nodes"]
+        direction BT
+        node1[var]
+        node2[x]
+        node3[:]
+        node4[i32]
+        node5[=]
+        node6[y]
+        node7[+]
+        node8[1]
+        node9[;]:::moved
+    end
+
+    classDef hidden visibility:hidden,display:none
+    classDef moved fill:#0F0,color:#000
+
+    node1 ~~~~ root
+    node3 --- node2 & node4
+    node3 ~~~ root
+    node5 ~~~~ root
+    node7 --- node6 & node8
+    node7 ~~~ root
+    node9 --- node1 & node3 & node5 & node7
+    node9 ~~~ root
+```
+
+Thus we have the parse tree:
+
+```mermaid
+flowchart BT
+    root:::hidden
+    subgraph nodes["Parsed nodes"]
+        direction BT
+        node1[var]
+        node2[x]
+        node3[:]
+        node4[i32]
+        node5[=]
+        node6[y]
+        node7[+]
+        node8[1]
+        node9[;]
+    end
+
+    classDef hidden visibility:hidden,display:none
+
+    node1 ~~~~ root
+    node3 --- node2 & node4
+    node3 ~~~ root
+    node5 ~~~~ root
+    node7 --- node6 & node8
+    node7 ~~~ root
+    node9 --- node1 & node3 & node5 & node7
+    node9 ~~~ root
+```
+
+In storage, this tree will be flat and in postorder. Because the order hasn't
+changed much from the original code, we can do the reordering for postorder with
+a minimal number of nodes being delayed for later output: it will be linear with
+respect to the depth of the parse tree.
+
+```mermaid
```

Comment:
This figure especially suffers from the translation to Mermaid, because you lose the way the nodes line up across the three levels.

Reply:
Yes, the first and third lines line up. I don't know Mermaid, but from what little I've seen, I'm not optimistic that it can give us the level of visual control that we would need to fix these problems, which is why I suggested switching to SVG.